### PR TITLE
Break up long function in trait selection error reporting + clean up nearby code

### PR DESCRIPTION
**Pull request by bryangarza**
- Move blocks of code into their own functions
- Replace a few function argument types with their type aliases
- Create "AppendConstMessage" enum to replace a nested `Option`.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 1]
```diff
@@ -2195,6 +2167,51 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         report(normalized_impl_candidates, err)
     }
 
+    fn report_similar_impl_candidates_for_root_obligation(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        trait_predicate: ty::Binder<'tcx, ty::TraitPredicate<'tcx>>,
+        body_def_id: LocalDefId,
+        mut err: &mut Diagnostic,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2175]:
```suggestion
        err: &mut Diagnostic,
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -2195,6 +2167,51 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+            && !self.tcx.trait_is_auto(def_id)
+            && !self.tcx.lang_items().iter().any(|(_, id)| id == def_id)
+        {
+            let trait_ref = trait_pred.to_poly_trait_ref();
+            let impl_candidates = self.find_similar_impl_candidates(trait_pred);
+            self.report_similar_impl_candidates(
+                &impl_candidates,
+                trait_ref,
+                body_def_id,
+                &mut err,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2209]:
```suggestion
                err,
```

**Review by compiler-errors — APPROVED** [review]
Just noting that a title like "Break up long function in trait selection error reporting" suggest that this PR is doing more than just splitting out one helper function from error reporting, so I wouldn't consider that function's size problem to have been vanquished just yet.

But there's nothing necessarily wrong with this impl, so r=me

**Comment by compiler-errors** [comment]
One nit though

r? @compiler-errors

**Comment by bryangarza** [comment]
> Just noting that a title like "Break up long function in trait selection error reporting" suggest that this PR is doing more than just splitting out one helper function from error reporting, so I wouldn't consider that function's size problem to have been vanquished just yet.

Good point, I’ll rename it to be more specific. I do plan on putting up more changes (or maybe I should do it all in one go in this PR).

**Comment by compiler-errors** [comment]
> I do plan on putting up more changes (or maybe I should do it all in one go in this PR).

No preference I guess. If you do plan on working more on this PR, I'd mark this one as waiting-on-author just to make it clear it's not looking for a review yet.

**Comment by bryangarza** [comment]
> > I do plan on putting up more changes (or maybe I should do it all in one go in this PR).
> 
> No preference I guess. If you do plan on working more on this PR, I'd mark this one as waiting-on-author just to make it clear it's not looking for a review yet.

Sounds good, I’ll just update this same PR tomorrow then

@rustbot label +S-waiting-on-author

**Comment by bryangarza** [comment]
Okay, I split it up a lot more :) I tried to keep most of the `if` checks in the main function itself, but move the code that inserts the errors/notes etc as their own functions, that way the logic can still be read top to bottom.

**Review by compiler-errors — APPROVED** [review]
Maybe this could use some additional simplification but this is a great start. Thanks!

**Pull request merged**